### DISCUSSION TITLE
Bluetooth?

### DISCUSSION THREAD

**Discussion by JPhilC** [Tigra-Astronomy/TA.ReactiveCommunications#17] [Q&A]
Hi Tim,

I am currently working on a small project to allow a mobile phone to be used to control some aspects of an ASCOM telescope driver. I have basic Rfcomm working between a Xamarin Forms Android app and a WPF Windows 10 desktop application in a test project using the 32Feet library for InTheHand. My next step is to make the communications asynchronous and two way. When a bluetooth connection is made it contains two streams (send and receive). What I am thinking of doing is attempting to extend the ReactiveCommunications library to wrap the bluetooth channel. Before I start, I thought I would ask if you believe this would be possible or would I be wasting my time.

Thanks,

Phil

**Comment by NameOfTheDragon** [comment]
> I thought I would ask if you believe this would be possible or would I be wasting my time.

Thanks for looking at RxComms. I'm probably a bit late responding to this and I've only just seen it.

All of the implementations I've done have been between the driver and the device, but I don't see any reason why this wouldn't work on the driver side. From the driver's perspective, if you treat your Bluetooth connection as a ICommunicationsChannel then you can use what I call the *Asynchronous Model* to parse your input stream and hook it up to the right code. This seems not completely different to how I added the hardware simulator into the Digital Domeworks driver. In that situation, I wrapped all the simulation code in a custom implementation of ICommunicationsChannel, called SimulatorCommunicationsChannel. In your case though, you might even get away with just using the built-in serial implementation if your Bluetooth connection can be made to "look like" a serial port.

Another approach to consider is to make your ASCOM driver support the Alpaca interface, which gives it a REST API, in effect. You can then have your control app work over WiFi or cellular networks. You could also pass the Alpaca requests via Bluetooth for times when there's no network available but you might have to implement some sort of rudimentary network layer, possibly with some help from RxComms on the driver side, to help with handling the requests.

It's a broad question and sometimes the devil is in the details, so it's hard to give concrete answers. RxComms is very flexible if you can bend your mind into the right paradigm so I have no doubt it could be made to work.

Hope that helps,
Tim
